README.md
# justino-ig-fetcher (Under development)

Scrape data from Instagram without applying for the authenticated API.

## Getting started

### Prerequisites

- NodeJS
- NPM
- Yarn

### Install

From [npm](https://www.npmjs.com/package/justino-ig-fetcher)

`yarn add justino-ig-fetcher`

or

`npm i justino-ig-fetcher --save`

### Test

- `git clone https://github.com/rafaeljustinox/justino-ig-fetcher.git`
- `yarn install` or `npm install`
- `yarn test` or `npm run test`

Optional environment variables for more complete testing :
- `SESSION_ID` : a session ID for authentication test and authenticated tests
- `PUBLIC_PROFILE` : a public profile to access
- `PRIVATE_PROFILE` : a private profile to access
- `STORY_PROFILE_ID` : a profile ID with a story to read
- `STORY_PROFILE_USERNAME` : a profile username with a story to read
- `HASHTAG` (default value : `cat`) : a hashtag to fetch
- `LOCATION_ID` (default value : `6889842` aka. Paris) : a location to fetch
- `POST` : a post to fetch
- `SEARCH_PROFILE` : a profile to search for
- `SEARCH_HASHTAG` (default value : `cats`) : a hashtag to search for
- `SEARCH_LOCATION` (default value : `Paris`) : a location to search for

Methods not covered by tests :
- `subscribeUserPosts`
- `subscribeHashtagPosts`
- `subscribeAccountNotifications`

### Usage

```js
const Insta = require('justino-ig-fetcher');
const InstaClient = new Insta();
```

#### Authentication

Authentication allows you to access private profile as long as you follow them.

##### Importing your session ID

- Go to instagram.com
- Login *(if not already logged in)*
- Open development tools *(`Ctrl` + `Shift` + `I`)*
- Get the `sessionid` cookie value
	- For chromium-based browsers : `application` tab
	- For firefox-based browsers : `storage` tab

##### Code

```js
InstaClient.authBySessionId(yourSessionId)
	.then(account => console.log(account))
	.catch(err => console.error(err));
```

If authentication is successfull, you'll get the form data from `accounts/edit` :

```json
{
	"first_name": "",
	"last_name": "",
	"email": "",
	"is_email_confirmed": true,
	"is_phone_confirmed": true,
	"username": "",
	"phone_number": "",
	"gender": 1,
	"birthday": null,
	"biography": "",
	"external_url": "",
	"chaining_enabled": true,
	"presence_disabled": false,
	"business_account": false,
	"usertag_review_enabled": false
}
```

If your session ID is invalid, you'll get the `401` error.

*Username/password authentication may be supported in the future.*

#### Get

These methods allows you to get specific elements from Instagram while you know exactly what you're looking for.

##### Errors handling

`get` may return errors in the two following cases.

- Request error : failed to get data from Instagram (HTTP code)
- Parsing error : failed to parse data returned by Instagram (`406`)
- No content : nothing to parse (`204`)
- Authentication required : session ID required to access data (`401`)
- Too many requests : rate limit exceeded (`429`)
- Conflict : automation detected, password reset required (`409`)

##### Get profile by username

```js
InstaClient.getProfile(username)
	.then(profile => console.log(profile))
	.catch(err => console.error(err));
```

Result

- `id` *string* - Instagram identifier, only used for stories
- `name` *string* - public full [name](https://help.instagram.com/583107688369069)
- `pic` *url* - public profile [picture](https://help.instagram.com/557544397610546)
- `bio` *string* - public biography
<br>`website` *url* - public website
<br>[more info about bio & website](https://help.instagram.com/362497417173378)
- `private` *boolean* - account [private state](https://help.instagram.com/448523408565555)
- `access` *boolean* - access to the profile's feed
<br>In order to have access to a private account's feed, you must have sent him a follow request that he accepted.
- `verified` *boolean* - account [verified state](https://help.instagram.com/854227311295302)
- `followers` *integer* - number of users following this profile
- `following` *integer* - number of users this profile follows
- `posts` *integer* - number of posts this profile published
- `lastPosts` *array of posts* - last posts
<br>This property is empty (`[]`) when the profile doesn't have any post but `null` if `access` is `false` (denied).
- `link` *url* - link to the profile's page
- `business` *string* - business category (when applicable and profile unblocked)
- `user` *object* - user relevant properties **(while authenticated)** :
	- `mutualFollowers` *array of usernames* - people following you and this profile
	- `blocking` *boolean* - you blocked this profile
	- `blocked` *boolean* - this profile blocked you (only available property in `user` while `true`)
	- `requesting` *boolean* - you sent a follow request to this profile (if private)
	- `requested` *boolean* - this profile sent you a follow request (if yours is private)
	- `following` *boolean* - you're following this profile
	- `followed` *boolean* - this profile follows you

##### Get profile story (requires authentication)

##### Using profile ID

```js
InstaClient.getProfileStoryById(id)
    .then(profile => console.log(profile))
    	.catch(err => console.error(err));
```

##### Using profile username (will automatically request profile ID)

```js
InstaClient.getProfileStory(username)
    .then(profile => console.log(profile))
    	.catch(err => console.error(err));
```

##### Result

- `unread` *boolean* - profile story is unread
- `author` *object* - a subset of profile
    - `username`
    - `pic`
- `user` *object* - user relevant properties
    - `requesting`
    - `following`
- `items` *array of stories* - profile stories
    - `url` *string* - link to original story file (`jpg`, `mp4`, ...)
    - `type` *string* - story type : `photo` or `video`
    - `timestamp` *epoch*
    - `expirationTimestamp` *epoch*

*Those methods will return `null` when a profile has no story.*

Note : calling this method will not mark the story as read.

##### Get hashtag

```js
InstaClient.getHashtag(hashtag)
	.then(hashtag => console.log(hashtag))
	.catch(err => console.error(err));
```

Result

- `pic` *url* - hashtag profile pic (can't find out how it is chosen)
- `posts` *integer* - number of posts containing this hashtag
- `featuredPosts` *array of posts* - featured posts published with this hashtag
<br>`lastPosts` *array of posts* - last posts published with this hashtag
<br>[more info about hashtag posts](https://help.instagram.com/777754038986618)
- `link` *url* - link to the hashtag's page
- `user` *object* - user relevant properties **(while authenticated)** :
	- `following` *boolean* - you [subscribed](https://help.instagram.com/2003408499915301) to this hashtag (receiving posts in your personal feed)

##### Get location by ID

Unfortunately, using IDs is currently the only way to get a location, at least for now.

```js
InstaClient.getLocation(id)
	.then(location => console.log(location))
	.catch(err => console.error(err));
```

Result

- `pic` *url* - location profile pic
- `posts` *integer* - posts published from that location
- `address` *object*
	- `street` *string*
	- `zipCode` *string*
	- `city` *string*
	- `latitude` *float*
	- `longitude` *float*
- `website` *url* - place's website
- `phone` *string* - place's contact phone number
- `featuredPosts` *array of posts* - featured posts published from this location
<br>`lastPosts` *array of posts* - last posts published from this location
- `link` *url* - link to this location's page

##### Array of posts

This is a subset of a real post, containing the following properties :

- `shortcode` *string* - post identifier
- `caption` *string* - post description
- `comments` *integer* - number of comments
- `likes` *integer* - number of likes
- `thumbnail` *url* - post thumbnail
<br>Always static image wether it's a photo or a video post, lower quality.

##### Get post by shortcode

The shortcode is the post's identifier : the link to a post is instagram.com/p/shortcode.

```js
InstaClient.getPost(shortcode)
	.then(post => console.log(post))
	.catch(err => console.error(err));
```

Result

- `author` *object* - a subset of a profile's properties.
	- `username` *string*
	- `name` *string*
	- `pic` *url*
	- `verified` *boolean*
	- `link` *url*
- `location`
	- `name` *string*
	- `city` *string*
- `contents` *array of posts*
	- `type` *string* - post type : `photo` or `video`
	- `url` *string* - link to original post file (`jpg`, `mp4`, ...)
	- if `type` is `video` :
	<br>`thumbnail` *string* - link to thumbnail
	<br>`views` *integer* - number of views
- `tagged` *array of usernames* - people tagged in post contents
- `likes` *integer* - number of likes
- `caption` *string* - post description
- `hashtags` *array of hashtags* - hashtags mentioned in post description
- `mentions` *array of usernames* - people mentioned in post description
- `edited` *boolean* - caption edited
- `comments` *array of objects* (Max 40)
	- `user` *string* - comment author's username
	- `content` *string* - comment content
	- `timestamp` *epoch*
	- `hashtags` *array of hashtags*
	- `mentions` *array of usernames*
	- `likes` *integer*
- `commentCount` *integer*
- `timestamp` *epoch*
- `link` *string* - link to the post

#### Paginated getters (require authentication)

Paginated getters allows bulk data downloads.

Params :
- `maxCount` *integer* - max number of items to return
- `pageId` *string* (optional) - page navigation identifier

Result : array + `nextPageId` property

Sample :
```js
(async () => {
	const page0 = await somePaginatedGetter(someId, 50);
	const page1 = await somePaginatedGetter(someId, 50, page0.nextPageId);
	const page2 = await somePaginatedGetter(someId, 50, page1.nextPageId);
})();
```

The `pageId`/`nextPageId` property may contain a string of digits, a base64 string, or a JSON string, but always must be leaved untouched.

##### Get profile posts

Result in array : full post object

###### Using profile ID

```js
InstaClient.getProfilePostsById(profileId, maxCount, pageId)
    .then(posts => console.log(posts))
    	.catch(err => console.error(err));
```

###### Using profile username (will automatically request profile ID)

```js
InstaClient.getProfilePosts(profileUsername, maxCount, pageId)
    .then(posts => console.log(posts))
    	.catch(err => console.error(err));
```

##### Get post comments

```js
InstaClient.getPostComments(shortcode, maxCount, pageId)
	.then(posts => console.log(posts))
	.catch(err => console.error(err));
```

Result in array : comment object

##### Get hashtag posts

```js
InstaClient.getHashtagPosts(hashtag, maxCount, pageId)
	.then(posts => console.log(posts))
	.catch(err => console.error(err));
```

Result in array : partial post object

##### Get location posts

```js
InstaClient.getLocationPostsById(locationId, maxCount, pageId)
	.then(posts => console.log(posts))
	.catch(err => console.error(err));
```

Result in array : partial post object

#### Search

##### Search profile

```js
InstaClient.searchProfile(query)
	.then(profiles => console.log(profiles))
	.catch(err => console.error(err));
```

Result in array : a subset of profile.

- `username`
- `name`
- `pic`
- `private`
- `verified`
- `followers`
- `user`
	- `following`

##### Search hashtag

```js
InstaClient.searchHashtag(hashtag)
	.then(hashtags => console.log(hashtags))
	.catch(err => console.error(err));
```

Result in array : a subset of hashtag.

- `name`
- `posts`

##### Search location

```js
InstaClient.searchLocation(location)
	.then(locations => console.log(locations))
	.catch(err => console.error(err));
```

Result in array : a subset of location.

- `id`
- `name`
- `address`
	- `street`
	- `city`
	- `latitude`
	- `longitude`

#### Subscribe to posts

- `options` *object* (optional)
    - `interval` *integer* (optional) - time in seconds between requests. **Default : 30**
    - `lastPostShortcode` *string* (optional) - shortcode from which to begin if not the next one to be published.
    - `fullPosts` *boolean* (optional) - fetch full post data, additional request required

##### From user

```js
InstaClient.subscribeUserPosts(username, (post, err) => {
    if(post)
        console.log(post.shortcode);
    else
        console.error(err);
}, {
    interval,
    lastPostShortcode,
    fullPosts
});
```

##### From hashtag

```js
InstaClient.subscribeHashtagPosts(hashtag, (post, err) => {
    if(post)
        console.log(post.shortcode);
    else
        console.error(err);
}, {
    interval,
    lastPostShortcode,
    fullPosts
});
```

#### Account requests (user-relevant methods)

##### Get account notifications

```js
InstaClient.getAccountNotifications()
	.then(notifications => console.log(notifications))
	.catch(err => console.error(err));
```

Result in array : notification

- `id` *string* - Notification identifier
- `timestamp` *epoch*
- `type` *string* - Notification type : `like`, `mention`, `comment`, `follow`
- `post`
	- `shortcode`
	- `thumbnail`
- `by`
	- `username`
	- `name`
	- `pic`
- `content` *string* - Comment content (when applicable)

##### Subscribe to account notifications

- `options` *object* (optional)
    - `interval` *integer* (optional) - time in seconds between requests. **Default : 30**
    - `lastNotificationId` *string* (optional) - Notification ID

```js
InstaClient.subscribeAccountNotifications((post, err) => {
    if(post)
        console.log(post.shortcode);
    else
        console.error(err);
}, {
    interval,
    lastNotificationId
});
```

##### Get account stories

```js
InstaClient.getAccountStories()
    .then(stories => console.log(stories))
    .catch(err => console.error(err));
```

Result in array : inbox-like

- `unread`
- `author` *object* - a subset of a profile's properties.
    - `id`
    - `username`
    - `pic`
- `user` *object* - user relevant properties
    - `requesting`
    - `following`

## Changelog

* `0.0.2` (2022-01-21) • Initial release

CHANGELOG.md
0.0.1 - First version

index.js
'use strict';

/*
Modules
 */

const
	https = require('https'),
	insta = 'https://www.instagram.com/';

/*
Utils
 */

const querystring = object => Object.keys(object).map(key => `${key}=${object[key]}`).join('&');

/*
Class private methods
 */

const self = {
	get: (path, sessionId, tryParse = true, params) => new Promise((resolve, reject) => {
		params = JSON.stringify({ __a: sessionId ? '1' : undefined, ...params });
		const url = insta + path + ((params !== '{}') ? ('/?' + querystring(JSON.parse(params))) : (tryParse ? '/' : ''));
		https.get(url, {
			headers: {
				cookie: sessionId ? `sessionid=${sessionId}` : ''
			}
		}, res => {
			let body = '';
			res.on('data', chunk => body += chunk);
			res.on('end', () => {
				if(res.statusCode !== 200){
					switch(res.statusCode){
						case 302: {
							switch(res.headers.location){
								case insta + 'accounts/login/':
									return reject(429);
								case insta + 'accounts/login/?next=/accounts/edit/%3F__a%3D1':
									return reject(401);
								default: {
									if(res.headers.location.startsWith(insta + 'challenge/?next='))
										return reject(409);
									reject(res.statusCode);
								}
							}
							break;
						}
						default: reject(res.statusCode);
					}
				}
				else if(tryParse){
					try {
						resolve(Object.values(JSON.parse(body)['graphql'] || JSON.parse(body))[0]);
					}
					catch(_){
						try {
							resolve(Object.values(Object.values(JSON.parse(body.match(/_sharedData = (.+);/)[1])['entry_data'])[0][0]['graphql'])[0]);
						}
						catch(_){
							reject(406);
						}
					}
				}
				else {
					resolve(body);
				}
			});
			res.on('error', reject);
		});
	}),
	search: (query, sessionId) => new Promise((resolve, reject) => self.get('web/search/topsearch', sessionId, false, { context: 'blended', query })
		.then(body => resolve(JSON.parse(body)))
		.catch(reject)),
	graphQL: (query, queryHash, sessionId) => self.get('graphql/query', sessionId, undefined, {
		query_hash: queryHash,
		variables: query ? JSON.stringify(query) : undefined
	}),
	partialPost: post => ({
		shortcode: post['node']['shortcode'],
		caption: post['node']['edge_media_to_caption']['edges'].length > 0
			? post['node']['edge_media_to_caption']['edges'][0]['node']['text'] : null,
		comments: post['node']['edge_media_to_comment']['count'],
		likes: post['node']['edge_liked_by']['count'],
		thumbnail: post['node']['display_url'],
		timestamp: post['node']['taken_at_timestamp']
	}),
	hashtagsRegex: /(?<=[\s>])#(\d*[A-Za-z_]+\d*)\b(?!;)/g,
	usernamesRegex: /@([A-Za-z0-9_](?:(?:[A-Za-z0-9_]|(?:\\.(?!\\.))){0,28}(?:[A-Za-z0-9_]))?)/g,
	postComment: comment => ({
		id: comment['node']['id'],
		user: comment['node']['owner']['username'],
		content: comment['node']['text'],
		timestamp: comment['node']['created_at'],
		hashtags: comment['node']['text'].match(self.hashtagsRegex),
		mentions: comment['node']['text'].match(self.usernamesRegex),
		likes: comment['node']['edge_liked_by']['count']
	}),
  partialReels: reels => {
    const reelsVideo = reels[0];

    const versions = reelsVideo.video_versions;
    // Sorting by quality
    versions.sort((a, b) => b.width - a.width);
    const bestQualityVideo = versions[0];
    const url = bestQualityVideo.url;
    return {
      shortcode_media: url
    };
  },
	fullPost: post => {
		const
			caption = post['edge_media_to_caption']['edges'].length > 0
				? post['edge_media_to_caption']['edges'][0]['node']['text'] : null,
			username = post['owner']['username'],
			shortcode = post['shortcode'];
		return {
			shortcode,
			author: {
				id: post['owner']['id'],
				username,
				name: post['owner']['full_name'],
				pic: post['owner']['profile_pic_url'],
				verified: post['owner']['is_verified'],
				link: `${ insta }/${ username }`
			},
			location: post['location'] ? {
				id: post['location']['id'],
				name: post['location']['name'],
				...(post['location']['address_json'] ? {
					city: JSON.parse(post['location']['address_json'])['city_name']
				} : {})
			} : null,
			...(post['__typename'] === 'GraphImage' ? {
				contents: [{
					type: 'photo',
					url: post['display_url']
				}]
			} : {}),
			...(post['__typename'] === 'GraphVideo' ? {
				contents: [{
					type: 'video',
					url: post['video_url'],
					thumbnail: post['display_url'],
					views: post['video_view_count']
				}]
			} : {}),
			...(post['__typename'] === 'GraphSidecar' ? {
				contents: post['edge_sidecar_to_children']['edges']
					.map(content => ({
						type: content['node']['is_video'] ? 'video' : 'photo',
						url: content['node']['is_video'] ? content['node']['video_url'] : content['node']['display_url'],
						...(content['node']['is_video'] ? {
							thumbnail: content['node']['display_url'],
							views: content['node']['video_view_count']
						} : {})
					}))
			} : {}),
			...(post['edge_media_to_tagged_user'] ? {
				tagged: post['edge_media_to_tagged_user']['edges']
					.map(u => u['node']['user']['username'])
			} : {}),
			likes: post['edge_media_preview_like']['count'],
			caption,
			hashtags: caption ? caption.match(self.hashtagsRegex) : null,
			mentions: caption ? caption.match(self.usernamesRegex) : null,
			edited: post['caption_is_edited'] || false,
			...(post['edge_media_preview_comment'] ? {
				comments: post['comments_disabled'] ? null : post['edge_media_preview_comment']['edges'].map(self.postComment),
				commentCount: post['edge_media_preview_comment']['count']
			} : {}),
			timestamp: post['taken_at_timestamp'],
			link: insta + 'p/' + shortcode
		}
	}
};

/*
Memoization
 */
const profileIds = {};

/*
Class public properties & methods
 */

module.exports = class Insta {
	constructor(){
		this.sessionId = undefined;
		this.username = undefined;
		this.queryHashs = {};
	}
	authBySessionId(sessionId){
		return new Promise((resolve, reject) => self.get('accounts/edit', sessionId)
			.then(body => {
				if(this.sessionId)
					process.emitWarning('Session ID changed');
				this.sessionId = sessionId;
				this.username = body['username'];
				resolve(body);
			})
			.catch(reject));
	}
	getAccountNotifications(){
		return new Promise((resolve, reject) => {
			if(!this.sessionId) return reject(401);
			self.get('accounts/activity', this.sessionId).then(res => {
				resolve(res['activity_feed']['edge_web_activity_feed']['edges'].map(item => item['node']).map(notification => ({
					id: notification['id'],
					timestamp: notification['timestamp'],
					type: ({
						'GraphLikeAggregatedStory' : 'like',
						'GraphMentionStory': 'mention',
						'GraphCommentMediaStory': 'comment',
						'GraphFollowAggregatedStory': 'follow'
					})[notification['__typename']],
					...(notification['media'] ? {
						post: {
							shortcode: notification['media']['shortcode'],
							thumbnail: notification['media']['thumbnail_src']
						}
					} : {}),
					...(notification['user'] ? {
						by: {
							username: notification['user']['username'],
							name: notification['user']['full_name'],
							pic: notification['user']['profile_pic_url']
						}
					} : {}),
					...(notification['__typename'] === 'GraphMentionStory' ? {
						content: notification['text']
					} : {})
				})));
			}).catch(reject);
		});
	}
	getAccountStories(){
		return new Promise((resolve, reject) => {
			if(!this.sessionId) return reject(401);
			self.get('', this.sessionId, false, { __a: undefined }).then(body => {
				self.graphQL(
					undefined,
					body.match(/<link rel="preload" href="\/graphql\/query\/\?query_hash=(.+)&amp;/)[1],
					this.sessionId
				).then(body => {
					resolve(body['user']['feed_reels_tray']['edge_reels_tray_to_reel']['edges'].map(item => ({
						unread: item['node']['latest_reel_media'] !== item['node']['seen'],
						author: {
							id: item['node']['user']['id'],
							username: item['node']['user']['username'],
							pic: item['node']['user']['profile_pic_url']
						},
						user: {
							requesting: item['node']['user']['requested_by_viewer'],
							following: item['node']['user']['followed_by_viewer']
						}
					})));
				}).catch(reject);
			}).catch(reject);
		});
	}
	getProfile(username = this.username, anonymous = false){
		return new Promise((resolve, reject) => self.get(username, anonymous ? null : this.sessionId)
			.then(profile => {
				const
					id = profile['id'],
					access = !profile['is_private'] || !!profile['followed_by_viewer'] || profile['username'] === this.username;
				profileIds[username] = id;
				resolve({
					id,
					name: profile['full_name'],
					pic: profile['profile_pic_url_hd'],
					bio: profile['biography'],
					private: profile['is_private'],
					access,
					verified: profile['is_verified'],
					website: profile['external_url'],
					followers: profile['edge_followed_by']['count'],
					following: profile['edge_follow']['count'],
					posts: profile['edge_owner_to_timeline_media']['count'],
					lastPosts: access ? profile['edge_owner_to_timeline_media']['edges'].map(post => self.partialPost(post)) : null,
					link: insta + profile['username'],
					...(profile['is_business_account'] ? {
						business: profile['business_category_name']
					} : {}),
					...(this.sessionId ? {
						user: {
							mutualFollowers: profile['edge_mutual_followed_by']['edges'].map(item => item['node']['username']),
							blocking: profile['blocked_by_viewer'],
							blocked: profile['has_blocked_viewer'],
							requesting: profile['requested_by_viewer'],
							requested: profile['has_requested_viewer'],
							following: profile['followed_by_viewer'],
							followed: profile['follows_viewer']
						}
					} : {})
				});
			})
			.catch(err => {
				if(err === 204){
					this.getProfile(username, true)
						.then(profile => resolve(Object.assign(profile, {
							user: { blocked: true }
						})))
						.catch(reject);
				}
				else
					reject(err);
			}));
	}
	async _getProfileId(username){
		if(!profileIds[username])
			profileIds[username] = (await this.getProfile(username)).id;
		return profileIds[username];
	}
	async _getQueryHashs(){
		if(JSON.stringify(this.queryHashs) !== '{}') return this.queryHashs;
		const
			{
				Consumer,
				ConsumerLibCommons,
				TagPageContainer,
				LocationPageContainer,
			} = Object.fromEntries([
				...(await self.get('', this.sessionId, false, { __a: undefined }))
					.matchAll(/static\/bundles\/.+?\/(.+?)\.js\/.+?\.js/g)
			].map(_ => _.reverse())),
			mainScriptBody = await self.get(Consumer, undefined, false),
			secondaryScriptBody = await self.get(ConsumerLibCommons, undefined, false),
			hashtagScriptBody = await self.get(TagPageContainer, undefined, false),
			locationScriptBody = await self.get(LocationPageContainer, undefined, false),
			localQueryIdRegex = /queryId:"([^"]+)"/;
		const [
			,
			[, comment],
			,
			[, post]
		] = [...mainScriptBody.matchAll(/queryId:"([^"]+)"/g)];
		this.queryHashs = {
			// story: mainScriptBody.match(/50,[a-zA-Z]="([^"]+)",/)[1],
			anyPost: mainScriptBody.match(/RETRY_TEXT.+var [a-zA-Z]="([^"]+)",/)[1],
			post: secondaryScriptBody.match(/queryId:"([^"]+)"/)[1],
			comment,
			hashtag: hashtagScriptBody.match(localQueryIdRegex)[1],
			location: locationScriptBody.match(localQueryIdRegex)[1]
		};
		return this.queryHashs;
	}
	getProfileStoryById(id){
		return new Promise((resolve, reject) => {
			if(!this.sessionId) return reject(401);
			this._getQueryHashs().then(queryHashs => self.graphQL({
				reel_ids: [ id ],
				precomposed_overlay: false
			}, queryHashs.story, this.sessionId).then(data => resolve(data['reels_media'][0] ? {
				unread: data['reels_media'][0]['latest_reel_media'] !== data['reels_media'][0]['seen'],
				author: {
					username: data['reels_media'][0]['user']['username'],
					pic: data['reels_media'][0]['user']['profile_pic_url']
				},
				user: {
					requesting: data['reels_media'][0]['user']['requested_by_viewer'],
					following: data['reels_media'][0]['user']['followed_by_viewer']
				},
				items: data['reels_media'][0]['items'].map(item => ({
					url: item['is_video'] ? item['video_resources'][0]['src'] : item['display_url'],
					type: item['is_video'] ? 'video' : 'photo',
					timestamp: item['taken_at_timestamp'],
					expirationTimestamp: item['expiring_at_timestamp'],
					...(item['story_cta_url'] ? {
						externalLink: item['story_cta_url']
					} : {})
				}))
			} : null)).catch(reject)).catch(reject);
		});
	}
	getProfileStory(username = this.username){
		return new Promise((resolve, reject) => {
			this._getProfileId(username)
				.then(id =>
					this.getProfileStoryById(id)
						.then(resolve)
						.catch(reject))
				.catch(reject);
		});
	}
	async getProfilePostsById(profileId, maxCount, pageId){
		const res = await self.graphQL({
			id: profileId,
			first: maxCount,
			after: pageId
		}, (await this._getQueryHashs()).post, this.sessionId);
		return Object.assign(
			res['user']['edge_owner_to_timeline_media']['edges'].map(item => self.fullPost(item['node'])),
			{
				nextPageId: res['user']['edge_owner_to_timeline_media']['page_info']['has_next_page']
					? res['user']['edge_owner_to_timeline_media']['page_info']['end_cursor']
					: undefined
			}
		);
	}
	async getProfilePosts(profileUsername, maxCount, pageId){
		return this.getProfilePostsById(await this._getProfileId(profileUsername), maxCount, pageId);
	}
	getHashtag(hashtag){
		return new Promise((resolve, reject) => {
			const path = `explore/tags/${hashtag}`;
			self.get(path, this.sessionId)
				.then(hashtag => resolve({
					pic: hashtag['profile_pic_url'],
					posts: hashtag['edge_hashtag_to_media']['count'],
					featuredPosts: hashtag['edge_hashtag_to_top_posts']['edges'].map(post => self.partialPost(post)),
					lastPosts: hashtag['edge_hashtag_to_media']['edges'].map(post => self.partialPost(post)),
					link: insta + path,
					...(this.sessionId ? {
						user: {
							following: hashtag['is_following']
						}
					} : {})
				}))
				.catch(reject);
		});
	}
	async getHashtagPosts(hashtag, maxCount, pageId){
		const res = await self.graphQL({
			tag_name: hashtag,
			first: maxCount,
			after: pageId
		}, (await this._getQueryHashs()).hashtag, this.sessionId);
		return Object.assign(
			res['hashtag']['edge_hashtag_to_media']['edges'].map(post => self.partialPost(post)),
			{
				nextPageId: res['hashtag']['edge_hashtag_to_media']['page_info']['has_next_page']
					? res['hashtag']['edge_hashtag_to_media']['page_info']['end_cursor']
					: undefined
			}
		);
	}
	getLocation(id){
		return new Promise((resolve, reject) => {
			const path = `explore/locations/${id}`;
			self.get(path)
				.then(location => {
					const address = JSON.parse(location['address_json']);
					resolve({
						pic: location['profile_pic_url'],
						posts: location['edge_location_to_media']['count'],
						address: {
							street: address['street_address'],
							zipCode: address['zip_code'],
							city: address['city_name'],
							latitude: location['lat'],
							longitude: location['lng']
						},
						website: location['website'],
						phone: location['phone'],
						featuredPosts: location['edge_location_to_top_posts']['edges'].map(post => self.partialPost(post)),
						lastPosts: location['edge_location_to_media']['edges'].map(post => self.partialPost(post)),
						link: insta + path
					});
				})
				.catch(reject);
		});
	}
	async getLocationPostsById(locationId, maxCount, pageId){
		const res = await self.graphQL({
			id: locationId,
			first: maxCount,
			after: pageId
		}, (await this._getQueryHashs()).location, this.sessionId);
		return Object.assign(
			res['location']['edge_location_to_media']['edges'].map(post => self.partialPost(post)),
			{
				nextPageId: res['location']['edge_location_to_media']['page_info']['has_next_page']
					? res['location']['edge_location_to_media']['page_info']['end_cursor']
					: undefined
			}
		);
	}
	getReels(shortcode, { useGraphQL = false } = {}){
		return new Promise((resolve, reject) => {
			if(useGraphQL){
				this._getQueryHashs().then(queryHashs => {
					self.graphQL({ shortcode }, queryHashs.anyPost, this.sessionId)
						.then(data => resolve(self.partialReels(data['shortcode_media'])))
						.catch(reject);
				});
			}
			else {
				self.get(`reel/${shortcode}`, this.sessionId)
					.then(post => resolve(self.partialReels(post)))
					.catch(reject);
			}
		});
	}
	getPost(shortcode, { useGraphQL = false } = {}){
		return new Promise((resolve, reject) => {
			if(useGraphQL){
				this._getQueryHashs().then(queryHashs => {
					self.graphQL({ shortcode }, queryHashs.anyPost, this.sessionId)
						.then(data => resolve(self.fullPost(data['shortcode_media'])))
						.catch(reject);
				});
			}
			else {
				self.get(`p/${shortcode}`, this.sessionId)
					.then(post => resolve(self.fullPost(post)))
					.catch(reject);
			}
		});
	}
	async getPostComments(shortcode, maxCount, pageId){
		const res = await self.graphQL({
			shortcode,
			first: maxCount,
			after: pageId
		}, (await this._getQueryHashs()).comment, this.sessionId);
		return Object.assign(
			res['shortcode_media']['edge_media_to_parent_comment']['edges'].map(self.postComment),
			{
				nextPageId: res['shortcode_media']['edge_media_to_parent_comment']['page_info']['has_next_page']
					? res['shortcode_media']['edge_media_to_parent_comment']['page_info']['end_cursor']
					: undefined
			}
		);
	}
	searchProfile(query){
		return new Promise((resolve, reject) => self.search(query, this.sessionId)
			.then(res => resolve(res['users'].map(item => item['user']).map(profile => ({
				username: profile['username'],
				name: profile['full_name'],
				pic: profile['profile_pic_url'],
				private: profile['is_private'],
				verified: profile['is_verified'],
				followers: profile['follower_count'],
				...(this.sessionId ? {
					user: {
						following: profile['following']
					}
				} : {})
			}))))
			.catch(reject));
	}
	searchHashtag(query){
		return new Promise((resolve, reject) => self.search(query)
			.then(res => resolve(res['hashtags'].map(item => item['hashtag'])
				.map(hashtag => ({ name: hashtag['name'], posts: hashtag['media_count'] }))))
			.catch(reject));
	}
	searchLocation(query){
		return new Promise((resolve, reject) => self.search(query)
			.then(res => resolve(res['places'].map(item => item['place']['location']).map(location => ({
				id: location['pk'],
				name: location['name'],
				address: {
					street: location['address'],
					city: location['city'],
					latitude: location['lat'],
					longitude: location['lng']
				}
			}))))
			.catch(reject));
	}
	subscribeAccountNotifications(callback, {
		interval = 30,
		lastNotificationId
	}){
		let active = true;
		const checkNewNotifications = () => {
			if(!active) return;
			(async () => {
				try {
					const notifications = await this.getAccountNotifications();
					const lastNotificationIndex = notifications.findIndex(notification => notification.id === lastNotificationId);
					if(lastNotificationIndex !== -1){
						for(let i = lastNotificationIndex - 1; i > -1 ; i--){
							callback(notifications[i]);
						}
					}
					lastNotificationId = notifications[0].id;
					setTimeout(checkNewNotifications, interval * 1000);
				}
				catch(err){
					callback(undefined, err);
					checkNewNotifications();
				}
			})();
		};
		checkNewNotifications();
		return {
			unsubscribe: () => {
				active = false;
			}
		};
	}
	subscribeUserPosts(username, callback, {
		interval = 30,
		lastPostShortcode,
		fullPosts = false
	} = {}){
		let active = true;
		const checkNewPosts = () => {
			if(!active) return;
			(async () => {
				try {
					const profile = await this.getProfile(username);
					const lastPostIndex = profile.lastPosts.findIndex(post => post.shortcode === lastPostShortcode);
					if(lastPostIndex !== -1){
						for(let i = lastPostIndex - 1; i > -1 ; i--){
							callback(fullPosts ? (await this.getPost(profile.lastPosts[i].shortcode)) : profile.lastPosts[i]);
						}
					}
					lastPostShortcode = profile.lastPosts[0].shortcode;
					setTimeout(checkNewPosts, interval * 1000);
				}
				catch(err){
					callback(undefined, err);
					checkNewPosts();
				}
			})();
		};
		checkNewPosts();
		return {
			unsubscribe: () => {
				active = false;
			}
		};
	}
	subscribeHashtagPosts(hashtagName, callback, {
		interval = 30,
		lastPostShortcode = undefined,
		fullPosts = false
	} = {}){
		let active = true;
		const checkNewPosts = () => {
			if(!active) return;
			(async () => {
				try {
					const hashtag = await this.getHashtag(hashtagName);
					const lastPostIndex = hashtag.lastPosts.findIndex(post => post.shortcode === lastPostShortcode);
					for(let i = lastPostIndex - 1; i > -1 ; i--){
						callback(fullPosts ? (await this.getPost(hashtag.lastPosts[i].shortcode)) : hashtag.lastPosts[i]);
					}
					lastPostShortcode = hashtag.lastPosts[0].shortcode;
					setTimeout(checkNewPosts, interval * 1000);
				}
				catch(err){
					callback(undefined, err);
					checkNewPosts();
				}
			})();
		};
		checkNewPosts();
		return {
			unsubscribe: () => {
				active = false;
			}
		};
	}
};